File: src/group-anagrams/group-anagrams.py
# https://leetcode.com/problems/group-anagrams

class Solution1(object):
    def char_map(self, s):
        d = dict()
        for char in s:
            if char in d:
                d[char] += 1
            else:
                d[char] = 1
        return d

    def serialize_char_map(self, char_map):
        s = ""
        for k in sorted(char_map):
            s += f"{k}{char_map[k]}"
        return s

    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        if strs == [""]:
            return [[""]]
        groups = dict()
        for s in strs:
            char_map = self.char_map(s)
            key = self.serialize_char_map(char_map)
            if key in groups:
                groups[key].append(s)
            else:
                groups[key] = [s]
        return list(groups.values())

# Just lean on python's inbuilt sorted from the original string, rather
# than sorting the char count map keys
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        groups = dict()
        for s in strs:
            key = "".join(sorted(s))
            if key in groups:
                groups[key].append(s)
            else:
                groups[key] = [s]
        return list(groups.values())

if __name__ == "__main__":
    sol = Solution()
    tests = [
        (
            ["eat","tea","tan","ate","nat","bat"],
            [["bat"],["nat","tan"],["ate","eat","tea"]],
        ),
        (
            [""],
            [[""]],
        ),
        (
            ["a"],
            [["a"]]
        )
    ]
    for (i, (strs, solution)) in enumerate(tests):
        result = sol.groupAnagrams(strs)
        solution, result = sorted(solution), sorted(result)
        # Visual inspection, unorderdness makes assert hard
        print(f"test {i} | result {result}, solution {solution}")
    print("✅ All tests passed")

File: src/group-anagrams/group-anagrams.rs
// https://leetcode.com/problems/group-anagrams
struct Solution;

use std::collections::HashMap;

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut map: HashMap<Vec<char>, Vec<String>> = HashMap::new();

        for s in strs {
            let mut v: Vec<char> = s.chars().collect();
            v.sort();
            map.entry(v).or_insert(vec![]).push(s);
        }

        map.into_iter().map(|(_k, v)| v).collect()
    }
}

fn main() {
    let tests = vec![(
        vec![
            "eat".to_string(),
            "tea".to_string(),
            "tan".to_string(),
            "ate".to_string(),
            "nat".to_string(),
            "bat".to_string(),
        ],
        vec![
            vec!["bat".to_string()],
            vec!["nat".to_string(), "tan".to_string()],
            vec!["ate".to_string(), "eat".to_string(), "tea".to_string()],
        ],
    )];

    // Writing a test assertion is a pain in the butt for this one, so we're just manually
    // inspecting the output and submitting it.

    for (strings, solution) in tests {
        let result = Solution::group_anagrams(strings.clone());
        println!("solution = {:?}", solution);
        println!("result = {:?}", result);
    }
}
